documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+```cs
+public readonly record struct ButtonEvent<TDevice, TButton>(in TDevice DeviceState, TButton Button);
+public readonly record struct ClickEvent<TDevice, TButton>(in TDevice DeviceState, TButton Button, Vector2 Position);
+public readonly record struct AxisEvent<TDevice, TAxis>(in TDevice DeviceState, int Index, TAxis OldValue, TAxis NewValue);
+public readonly record struct TextEvent<TDevice>(in TDevice DeviceState, string? OldText, string? NewText);
+```
+
+This is the part of this proposal that incorporates the ideas in Enhanced Input Events, and is why this proposal supersedes that one.
+
+One final point to note is that throughout the rest of the proposal the following type will be used:
+
+```cs
+public struct InputReadOnlyList<T> : IReadOnlyList<T>
+{
+    public InputReadOnlyList(IReadOnlyList<T> other);
+}
+```
+
+The Silk.NET team wishes to reserve the right to add more constructors to this type as it sees fit.
+
+This exists so that, should the Silk.NET choose to, we can optimize the lookup of elements while ensuring things like indexers are inlined and don't result in a virtual call if our implementation allows us to do so.
+
+**INFORMATIVE TEXT:** For example, for joystick and mouse buttons we could use a fixed-sized bit buffer where each bit represents an individual button: 1 for pressed, 0 for unpressed. But for something like keyboard input where there are a large amount of keys, we can't do that and will likely use `Memory<T>` instead.
+
+# Mouse Input
+
+As discussed earlier, the interface will be very simple.
+
+```cs
+public interface IMouse : IInputDevice
+{
+    MouseState State { get; }
+    ICursorConfiguration Cursor { get; }
+    void SetPosition(Vector2 pos);
+}
+```
+
+`State` is the device state as defined earlier.
+
+`Cursor` contains the cursor configuration. This isn't actually state that the end user can change, and has been made an interface rather than a state struct accordingly.
+
+`SetPosition` allows moving the mouse cursor without the end user physically moving their mouse.
+
+The device state returned by `State` fills out the following structure:
+
+```cs
+public struct MouseState
+{
+    public IMouse Device { get; init; }
+    public MouseButtonState Buttons { get; init; }
+    public ICursorConfiguration Cursor { get; } // forwards to Device.Cursor
+    public Vector2 Position { get; set; } // all sets after the first set forward to Device.SetPosition
+    public Vector2 WheelPosition { get; init; }
+}
+```
+
+For ease-of-use, all APIs on `Device` (other than state) are accessible via the state struct, as indicated in the comments in the API snippet.
+
+`MouseButtonState` is defined as:
+```cs
+public struct MouseButtonState
+{
+    public bool this[MouseButton btn] { get; }
```

Comment:
Why is this not init-able?

Reply:
This is addressed in the comment on line 281, it doesn't make sense for anything else being the case.